pkg/controllers/cspc-controller/status.go

```diff
@@ -0,0 +1,151 @@
+	cstor "github.com/openebs/api/pkg/apis/cstor/v1"
+	"github.com/openebs/api/pkg/apis/types"
+	"github.com/openebs/cstor-operators/pkg/controllers/cspc-controller/util"
+	"github.com/pkg/errors"
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"k8s.io/klog"
+	"time"
+)
+
+func (c *Controller) UpdateStatusEventually(cspc *cstor.CStorPoolCluster) error {
+	err := c.UpdateStatus(cspc)
+	maxRetry := 3
+
+	if err != nil {
+		klog.Errorf("failed to update cspc %s status: will retry %d times at 2s interval: {%s}",
+			cspc.Name, maxRetry, err.Error())
+		for maxRetry > 0 {
+			cspcNew, err := c.GetStoredCStorVersionClient().
+				CStorPoolClusters(cspc.Namespace).
+				Get(cspc.Name, metav1.GetOptions{})
+
+			if err != nil {
+				// this is possible due to etcd unavailability so do not retry more here
+				return errors.Wrapf(err, "failed to update cspc status")
+			}
+
+			err = c.UpdateStatus(cspcNew)
+			if err != nil {
+				maxRetry = maxRetry - 1
+				klog.Errorf("failed to update cspc %s status: will retry %d times at 2s interval : {%s}",
+					cspc.Name, maxRetry, err.Error())
+				time.Sleep(2 * time.Second)
+				continue
+			}
+			return nil
+		}
+
+	}
+	return err
+}
+
+func (c *Controller) UpdateStatus(cspc *cstor.CStorPoolCluster) error {
+	status, err := c.calculateStatus(cspc)
+	if err != nil {
+		return errors.Wrapf(err, "failed to calculate cspc %s status", cspc.Name)
+	}
+	cspc.Status = status
+	_, err = c.GetStoredCStorVersionClient().CStorPoolClusters(cspc.Namespace).Update(cspc)
+
+	if err != nil {
+		return errors.Wrapf(err, "failed to update cspc %s in namespace %s", cspc.Name, cspc.Namespace)
+	}
+
+	return nil
+}
+
+func (c *Controller) calculateStatus(cspc *cstor.CStorPoolCluster) (cstor.CStorPoolClusterStatus, error) {
+	var healthyCSPIs int32
+	cspiList, err := c.GetCSPIListForCSPC(cspc)
+	if err != nil {
+		return cstor.CStorPoolClusterStatus{}, errors.Wrapf(err, "failed to list cspi(s) for cspc %s in namespace %s", cspc.Name, cspc.Namespace)
+	}
+
+	// List all corresponding pool managers for the cspc
+	poolManagerList, err := c.kubeclientset.
+		AppsV1().
+		Deployments(cspc.Namespace).
+		List(metav1.ListOptions{LabelSelector: string(types.CStorPoolClusterLabelKey) + "=" + cspc.Name})
+	if err != nil {
+		return cstor.CStorPoolClusterStatus{}, errors.Wrapf(err, "failed to list pool-manager deployments for cspc %s in namespace %s", cspc.Name, cspc.Namespace)
+	}
+
+	cspiNameToPoolManager := make(map[string]appsv1.Deployment)
+
+	for _, poolmanager := range poolManagerList.Items {
+		poolmanager := poolmanager // pin it
+		// note: name of cspi and corresponding pool manager is same.
+		cspiNameToPoolManager[poolmanager.Name] = poolmanager
+	}
+
+	provisionedCSPIs := int32(len(cspiList.Items))
+	desiredCSPIs := int32(len(cspc.Spec.Pools))
+
+	status := cstor.CStorPoolClusterStatus{
+		ProvisionedInstances: provisionedCSPIs,
+		DesiredInstances:     desiredCSPIs,
+	}
```

Comment:
```go
status := cstor.CStorPoolClusterStatus{
		ProvisionedInstances: int32(len(cspiList.Items)),
		DesiredInstances:     int32(len(cspc.Spec.Pools)),
	}
```
Can we change like above , as we are not  using these 2 variables any where else

Reply:
sure